Signal/src/util/ThreadUtil.m

```diff
@@ -120,43 +123,162 @@ + (ThreadOffersAndIndicators *)ensureThreadOffersAndIndicators:(TSThread *)threa
+                              }
+
+                              TSInteraction *interaction = (TSInteraction *)object;
+
+                              if (interaction.timestampForSorting
+                                  < firstUnseenInteractionTimestamp.unsignedLongLongValue) {
+                                  // By default we want the unread indicator to appear just before
+                                  // the first unread message.
+                                  *stop = YES;
+                                  return;
+                              }
+
+                              visibleUnseenMessageCount++;
+
+                              interactionAfterUnreadIndicator = interaction;
+
+                              if (visibleUnseenMessageCount + 1 >= maxRangeSize) {
+                                  // If there are more unseen messages than can be displayed in the
+                                  // messages view, show the unread indicator at the top of the
+                                  // displayed messages.
+                                  *stop = YES;
+                                  hasMoreUnseenMessages = YES;
+                              }
+                          }];
+
+            OWSAssert(interactionAfterUnreadIndicator);
+
+            if (hasMoreUnseenMessages) {
+                NSMutableSet<NSData *> *missingUnseenSafetyNumberChanges = [NSMutableSet set];
+                for (TSInvalidIdentityKeyErrorMessage *safetyNumberChange in safetyNumberChanges) {
+                    BOOL isUnseen = safetyNumberChange.timestampForSorting
+                        >= firstUnseenInteractionTimestamp.unsignedLongLongValue;
+                    if (!isUnseen) {
+                        continue;
+                    }
+                    BOOL isMissing
+                        = safetyNumberChange.timestampForSorting < interactionAfterUnreadIndicator.timestampForSorting;
+                    if (!isMissing) {
+                        continue;
+                    }
+                    [missingUnseenSafetyNumberChanges addObject:safetyNumberChange.newIdentityKey];
+                }
+
+                missingUnseenSafetyNumberChangeCount = missingUnseenSafetyNumberChanges.count;
+            }
+        }
+        result.firstUnseenInteractionTimestamp = firstUnseenInteractionTimestamp;
+        if (hasMoreUnseenMessages) {
+            // The unread indicator is _before_ the last visible unseen message.
+            result.unreadIndicatorPosition = @(visibleUnseenMessageCount);
+        }
+
         TSMessage *firstMessage = firstIncomingMessage;
         if (!firstMessage
-            || (firstOutgoingMessage &&
-                   [[firstOutgoingMessage receiptDateForSorting] compare:[firstMessage receiptDateForSorting]]
-                       == NSOrderedAscending)) {
+            || (firstOutgoingMessage && [firstOutgoingMessage compareForSorting:firstMessage] == NSOrderedAscending)) {
```

Comment:
I was actually thinking of making an even more readable method `isBefore:` for this kind of assert but I resisted.